docs/specs/lcb-traits.md

```diff
@@ -0,0 +1,302 @@
+    end trait
+
+Traits can be derived from multiple supertraits.
+
+## Trait objects
+
+    public trait U
+        method type F(in me)
+    end trait
+
+    type T
+
+    implementation of U for T
+        method F(in me)
+        end method
+    end implementation
+
+    public handler G(in pParam as U)
+        pParam.F()
+    end handler
+
+A trait object is created whenever a concrete type `T` needs to be passed by
+trait `U` rather than directly as an `T`.  It packages the `T` value with
+information that identifies the methods of the implementation of `U` that was
+used.
+
+## Further work
+
+### Parameterized types
+
+Traits do not fully solve the problem of implementing `formatted as string`
+syntax with full generality and type safety.  There remains the problem of
+formatting polymorphic compound types such as `list`.  Consider the following
+naïve approach:
+
+    public implementation of CanFormatAsString for list
+        method FormatAsString(in me) as string
+            variable tString as string
+            put "(" into tString
+
+            variable tIter as any
+            repeat for each element tElement of me
+                variable tElement as CanFormatAsString
+                put tIter into tElement -- TYPE UNSAFE
+                put tElement formatted as string after tString
+            end repeat
+
+            put ")" after tString
+            return tString
+        end method
+    end implementation
+
+If the above method is compiled with strict type checking, then the marked line
+should fail to compile because `tIter` is of unknown type `any`, and therefore
+does not necessarily implement the `CanFormatAsString` trait.
+
+It would be better if the `CanFormatAsString` trait was only defined for lists
+where every element also conforms to `CanFormatAsString`.
+
+Similar arguments apply for the `array` type.
+
+To get full generality and type safety of the `formatted as string` trait
+Builder therefore also needs type specialization (e.g. `list of
+CanFormatAsString`).
+
+### Interaction with runtime typing
+
+    trait type T
+    end trait
+
+    implementation of T for integer
+    end implementation
+
+    handler Confuse(in pIn as T) as T
+        return pIn
+    end handler
+
+    public handler main()
+        variable tVar as any
+        put 15 into tVar        -- contents of tVar is an integer
+        Confuse(tVar) into tVar -- contents of tVar is a T trait object
+        add 1 to tVar           -- error?
+    end handler
+
+It is not clear what the correct behaviour is here.
+
+Instead of an error, one option is to "unpack" any trait object when assigned
+to an `any` slot.  However, consider:
+
+    public trait type T
+        method F(in me)
+    end trait
+
+    public handler F(in pIn as T) -- pIn is a T trait object
+        variable tVar as any
+        put pIn into tVar         -- tVar is a ?? (out of lexical scope)
+
+        tVar.F()                  -- error?
+    end module
```

Comment:
It occurs to me that the distinct view and interface view are not incompatible - at a VM implementation level I think they come down to how fine-grained the operations are.

For the interface view, you just need an (Value as Trait).Method() operation.

For the distinct view, you need two operations 'Value as Trait' and 'Value.Method()' (the point here being that for 'Value as Trait' to actually have the desired effect it has to produce something which has a different runtime type attached to it than Value).